domain/src/main/java/org/oppia/domain/repository/SkillRepository.kt

```diff
@@ -0,0 +1,209 @@
+import androidx.lifecycle.MutableLiveData
+import kotlinx.coroutines.Deferred
+import javax.inject.Inject
+import javax.inject.Singleton
+import org.json.JSONObject
+import org.oppia.app.model.ConceptCardView
+import org.oppia.app.model.LessonThumbnail
+import org.oppia.app.model.LessonThumbnailGraphic
+import org.oppia.app.model.SkillSummaryDatabase
+import org.oppia.app.model.SkillSummaryDomain
+import org.oppia.data.persistence.PersistentCacheStore
+import org.oppia.domain.topic.FRACTIONS_SKILL_ID_0
+import org.oppia.domain.topic.FRACTIONS_SKILL_ID_1
+import org.oppia.domain.topic.FRACTIONS_SKILL_ID_2
+import org.oppia.domain.topic.FRACTIONS_TOPIC_ID
+import org.oppia.domain.topic.RATIOS_SKILL_ID_0
+import org.oppia.domain.topic.RATIOS_TOPIC_ID
+import org.oppia.domain.topic.TOPIC_FILE_ASSOCIATIONS
+import org.oppia.domain.util.JsonAssetRetriever
+import org.oppia.util.data.AsyncResult
+import org.oppia.util.data.DataProvider
+import org.oppia.util.data.DataProviders
+import org.oppia.util.logging.Logger
+
+private const val ADD_SKILL_SUMMARY_LIST_TRANSFORMED_PROVIDER_ID = "add_skill_list_transformed_id"
+private const val GET_SKILL_SUMMARY_TRANSFORMED_PROVIDER_ID = "get_skill_summary_transformed_id"
+
+/** Controller for retrieving all aspects of a topic. */
+@Singleton
+class SkillRepository @Inject constructor(
+  cacheStoreFactory: PersistentCacheStore.Factory,
+  private val dataProviders: DataProviders,
+  private val jsonAssetRetriever: JsonAssetRetriever,
+  private val logger: Logger
+) {
+
+  /** Indicates that the given skill summary already exists. */
+  class SkillSummaryAlreadyExistsException(msg: String) : Exception(msg)
+
+  /** Indicates that the given skill summary is not found. */
+  class SkillSummaryNotFoundException(msg: String) : Exception(msg)
+
+  /**
+   * These statuses correspond to the exceptions above such that if the deferred contains
+   * SKILL_SUMMARY_ALREADY_FOUND, the [SkillSummaryAlreadyExistsException] will be passed to a failed AsyncResult.
+   *
+   * SUCCESS corresponds to a successful AsyncResult.
+   */
+  private enum class SkillSummaryActionStatus {
+    SUCCESS,
+    SKILL_SUMMARY_ALREADY_EXISTS,
+    SKILL_SUMMARY_NOT_FOUND
+  }
+
+  private val skillSummaryDataStore =
+    cacheStoreFactory.create(
+      "skill_summary_database",
+      SkillSummaryDatabase.getDefaultInstance()
+    )
+
+  // TODO(#272): Remove init block when storeDataAsync is fixed
+  init {
+    skillSummaryDataStore.primeCacheAsync().invokeOnCompletion {
+      it?.let {
+        logger.e(
+          "DOMAIN",
+          "Failed to prime cache skillSummaryDataStore ahead of LiveData conversion for SkillRepository.",
+          it
+        )
+      }
+    }
+  }
+
+  /**
+   * This method stores all skills to proto database by reading these skills from corresponding json files
+   * and returns a successful [LiveData]
+   */
+  fun storeAllSkillSummariesToProtoDatabase(): LiveData<AsyncResult<Any?>> {
```

Comment:
This will be called at the start of the app, mainly in `SplashActivity` to save all skills to proto-database along with other contents like subtopics, topics, etc. which will be introduced later in different files.

Thats why this is kept public.

What exactly do you mean by `prime itself`? It should store all these information on its own without calling this function in SplashActivity? Not sure but is this similar to [TopicListController](https://github.com/oppia/oppia-android/blob/645cad0be4ed5d0536326c295c86319175e7cba3/domain/src/main/java/org/oppia/domain/topic/TopicListController.kt#L111) ?

Reply:
Well if it’s an async thing, we need to block other operations or the UI on it loading (where the former is preferred). I was originally thinking of TopicListController, but it would be more performant to load this lazily after one of the other methods are called (meaning all other methods combine their data provider with one that’s used to preload the cache).